Working towards using GitHub Actions for CI

andrewvaughanj opened the pull request:
First attempt at getting STP to build with GitHub Actions.

Some points:

- Travis is not yet disabled

- Moves towards using `contrib` style scripts (similar to Boolector and CVC4) for building STP's dependencies

- Submodules for Google Test and Output Check have been moved to `contrib` scripts

- Currently builds STP with CMS under both `gcc` and `clang`


Signed-off-by: Andrew V. Jones <andrew.jones@vector.com>

rgov commented:
Is this WIP or ready to be merged?

rgov commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,71 @@
+
+    - name: Packages
+      run: |
+        sudo mkdir -p /usr/share/man/man1
```
I'm curious why this is needed?

andrewvaughanj replied:
Hmm, I was testing stuff with https://github.com/nektos/act and hit some issue related to installing some dependancy, but then I changed the "base" image I was working and didn't remove this.

It can (probably) go -- I'll check!

rgov commented on contrib/setup-cms.sh:
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env bash
+
+set -e -o pipefail
```
Can you also add `-u` (expansion of unset variable fails)?

andrewvaughanj replied:
🤦 if someone had asked me, I would have been certain I had `-u`!

rgov commented on contrib/setup-cms.sh:
```diff
@@ -0,0 +1,21 @@
+
+dep="cms"
+
+cd ${dep_dir}
```
Please quote variable expansions. Maybe there's a shell linter we can apply to make sure it's all good.

andrewvaughanj replied:
Done.

Scripts have been checked with https://github.com/koalaman/shellcheck (scripts were already formatted in a way that https://github.com/mvdan/sh made no changes).

rgov commented on contrib/setup-cms.sh:
```diff
@@ -0,0 +1,21 @@
+git clone https://github.com/msoos/cryptominisat ${dep}
+cd ${dep}
+mkdir build && cd build
+cmake -G Ninja -DCMAKE_INSTALL_PREFIX:PATH=${install_dir} ..
```
Do we need to use the Ninja generator? It adds a prereq we don't need.

andrewvaughanj replied:
So, I actually removed Ninja *from the contrib scripts* as I agree that not everyone will have it.

I **have not** removed it from the GitHub Action.

msoos commented:
I don't think we need ninja, it works without it I'm sure. I never used
ninja to be honest so I don't know what it brings but in my understanding,
cmake and make should do the trick :) Maybe others know what the benefits
are?

On Tue, Aug 4, 2020, 04:48 Ryan Govostes <notifications@github.com> wrote:

> *@rgov* requested changes on this pull request.
> ------------------------------
>
> In .github/workflows/ci.yml
> <https://github.com/stp/stp/pull/378#discussion_r464766088>:
>
> > +
> +    name: ${{ matrix.name }}
> +    runs-on: [ubuntu-latest]
> +
> +    steps:
> +
> +    - uses: actions/checkout@v2
> +
> +    - name: Environment
> +      run: |
> +        echo "::set-env name=CC::${{ matrix.env.CC }}"
> +        echo "::set-env name=CXX::${{ matrix.env.CXX }}"
> +
> +    - name: Packages
> +      run: |
> +        sudo mkdir -p /usr/share/man/man1
>
> I'm curious why this is needed?
> ------------------------------
>
> In contrib/setup-cms.sh
> <https://github.com/stp/stp/pull/378#discussion_r464766384>:
>
> > @@ -0,0 +1,21 @@
> +#!/usr/bin/env bash
> +
> +set -e -o pipefail
>
> Can you also add -u (expansion of unset variable fails)?
> ------------------------------
>
> In contrib/setup-cms.sh
> <https://github.com/stp/stp/pull/378#discussion_r464766659>:
>
> > @@ -0,0 +1,21 @@
> +#!/usr/bin/env bash
> +
> +set -e -o pipefail
> +
> +dep_dir="deps"
> +install_dir=$(readlink -fm ${dep_dir}/install)
> +
> +[ ! -d "${install_dir}" ] && mkdir -p "${install_dir}"
> +
> +dep="cms"
> +
> +cd ${dep_dir}
>
> Please quote variable expansions. Maybe there's a shell linter we can
> apply to make sure it's all good.
> ------------------------------
>
> In contrib/setup-cms.sh
> <https://github.com/stp/stp/pull/378#discussion_r464766843>:
>
> > +#!/usr/bin/env bash
> +
> +set -e -o pipefail
> +
> +dep_dir="deps"
> +install_dir=$(readlink -fm ${dep_dir}/install)
> +
> +[ ! -d "${install_dir}" ] && mkdir -p "${install_dir}"
> +
> +dep="cms"
> +
> +cd ${dep_dir}
> +git clone https://github.com/msoos/cryptominisat ${dep}
> +cd ${dep}
> +mkdir build && cd build
> +cmake -G Ninja -DCMAKE_INSTALL_PREFIX:PATH=${install_dir} ..
>
> Do we need to use the Ninja generator? It adds a prereq we don't need.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/stp/stp/pull/378#pullrequestreview-460459121>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKF4ONB3FJ6RGFGGU5EJTLR65ZGPANCNFSM4PT2NZEQ>
> .
>

andrewvaughanj commented:
Ninja is just a “higher performance” generator that (in my opinion) does a better job of no-op builds.

It is used as the default generator for projects such as LLVM and Chrome (https://ninja-build.org). Plus, I *exclusively* build STP using Ninja (and any other project that works with it).

That being said, I’m fine to not use Ninja for this :)

msoos commented:
Ah then it's good! Again, I have no idea what it brings, so I'll trust you
on this :)

On Tue, Aug 4, 2020, 09:36 Andrew V. Jones <notifications@github.com> wrote:

> Ninja is just a “higher performance” generator that (in my opinion) does a
> better job of no-op builds.
>
> It is used as the default generator for projects such as LLVM and Chrome (
> https://ninja-build.org). Plus, I *exclusively* build STP using Ninja
> (and any other project that works with it).
>
> That being said, I’m fine to not use Ninja for this :)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/stp/stp/pull/378#issuecomment-668434318>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKF4OO5KHWYO767O4IFLO3R663APANCNFSM4PT2NZEQ>
> .
>

andrewvaughanj commented:
> Is this WIP or ready to be merged?

I think it is ready for a second round of review. Since the last time, I've:

- Removed Travis

- Updated the top-level README to tell people to use the `contrib` scripts

I think before we totally go ahead, we should make sure that Mate is happy that this action *does less* than Travis at the moment (no Coverity, no static builds, etc.).

Just so we're all clear, I have a number of follow-up bits I want to do this with:

- Add static builds

- Add builds with the other solvers (e.g., Riss)

- Migrate the Windows builds to use GitHub Actions

- Maybe we want to use https://docs.github.com/en/github/finding-security-vulnerabilities-and-errors-in-your-code/about-code-scanning over Coverity (I already have access)


Right now, this is baby steps towards that.

andrewvaughanj commented:
Scripts have been renamed from the "contrib" directory into the "scripts/deps" folder (and corresponding documentation updated).

msoos commented:
Wow, going along nicely!! Thanks for all the hard work :)

andrewvaughanj commented:
@msoos how does this look for you? It appears to be clean, and up to date with master!

msoos commented:
Sounds great to me! GitHub Actions are pretty fast and neat. I support this change for sure, thanks so much! :+1:

Hanh668 reviewed (commented):
> 
![7AE44F6B-2213-42BA-8608-921CFA74B9E8](https://user-images.githubusercontent.com/83554447/116826020-2f422f00-abbc-11eb-9286-3ae2f3efad64.jpeg)

rgov commented:
Thank you @andrewvaughanj. This was way overdue.

There are currently test case failures but it looks like CI is working OK. I also updated the CodeQL workflow to match the latest template.
